### ISSUE TITLE
Non-termination of runtime meta tasks during shutdown in -lg:resilient mode

### ISSUE THREAD

**Issue by karthikmurthy** [StanfordLegion/legion#359]
Short: Running any c++ example in legion/examples on the master/stable branch with -lg:resilient mode causes non-termination (e.g., ./ghost -ll:cpu 4 -lg:resilient). 

Preliminary debugging revealed: GDB attaching to the ghost c++ execution revealed that 224 meta tasks are waiting for the deferred_commit_trigger, cannot be committed, confirm_runtime_shutdown is not successful and hence, the shutdown meta task (taskID 4) keeps getting added/removed/added to the scheduler but not resulting in any progress. 

Discussion with @lightsighter : A solution should be that once the top level task completes, it should explicitly trigger the completion of at least/at most/just one of the meta tasks, and this should lead to all of them to commit (a solution that @streichler also proposed). In light of how things are with ghost example, we need to revisit this. 

DEBUG_SHUTDOWN_HANG: If enabled in such executions, causes a segfault. The reason appears to be that we force-commit/ignore the 224 meta tasks and end up proceeding to a further point where realm is trying to schedule a clean up task and fails. This is a sideeffect and the primary problem is the stated above.

Reproducers:  

02index_tasks fails with an assertion 'dependence_tracker.commit != null' 
07_partitioning faces non-termination, but 2 threads are stuck in dma, 2 threads in wait_for_work, 1 wait_for_shutdown
ghost (.tar.gz attached to this issue) is a reduced ghost example 

All apps from the tutorial available with master branch. No modifications necessary. 
$> export LG_RT_DIR=
$> make
$> ./app -lg:resilient

DETAILS for GHOST:
1) ./ghost -ll:cpu 2 -lg:resilient produces the hang (2 is sufficient)
2) It is a straightforward consumer and producer program, left and right ghost regions with access regulated via copy/acquire/release/init tasks. NO WORK is done here, i.e., stencil task is removed since it does not matter. 
3) If you replace acquire and release with copy ops, just for fun, things work fine. If you do only one iteration of the loop s, things work fine. If you do not do init_task (which is empty btw), then things run fine. 
4) In toto, we are looking at some weird interaction of init, copy, acquire, release interaction in the presence of phase barriers causing the problem.

[ghost.tar.gz](https://github.com/StanfordLegion/legion/files/1773113/ghost.tar.gz)

**Comment by lightsighter** [comment]
I have made some progress on this issue. There are three improvements so far:
- The crash from tutorial 02 should be gone. It actually runs correctly now.
- The hangs that we were seeing were actually hanging earlier than when the top-level task was completing, which was very bad. I've fixed this so that everything should now run all the way to completion including the top-level task, and then we will see the hangs. This should manifest as all tutorials reporting 'SUCCESS' but then failing to shutdown properly. If you see one behaving differently then please let me know.
- I've updated the 'gen' variable on objects so that is is incremented properly only when operations are actually committed and not earlier like we had been doing.

The issue of resolving the hangs on shutdown when operations are not committed is trickier. There are two problems to be worked through:
- First, we need to determine how to beginning triggering commits once we detect that the top-level task is complete (e.g. we will never need to rollback any of this task-tree). It seems like the way to do this is to walk down the task tree until we get to the leaf contexts and then begin triggering the completion of operations which have no commit dependences (e.g. the consumers at the end of the task that don't produce any output except maybe to their enclosing parent task). Once we get these tasks committing then the commits should flow first back through the context, and then as all the child operations are committed up through the task tree recursively exactly as we would expect.
- Second, what does it mean to commit an index space operation (task, fill, copy)? Is it committed when all its point operations are committed? How are we going to track hardening and verified on the region arguments? Will they be done on a per-point basis, or as an aggregated whole?
I have some thoughts on these issues, but we should discuss them in more detail before proceeding further as these decisions will have a bearing on the implementation.

**Comment by karthikmurthy** [comment]
The design changes needed to accommodate some of the issues raised are being developed in a separate branch (resilience). This issue is no longer relevant.
